doc/Language/haskell-to-p6.pod6

```diff
@@ -0,0 +1,420 @@
+
+    sub parse-int(Str $s --> Int:_) { ... }
+    
+    # One way to do it
+    given parse-int($string) {
+      when * ~~ Int:D { * }
+      when * ~~ Int:U { 0 }
+    }
+    
+    # Another way to do it
+    my Int $number = parse-int($string);
+    if $number.defined { $number } else { 0 }
+    
+    
+    # A better way
+    with parse-int($string) { * } else { 0 }
+    
+The `with` operator that you see above is like `if`, except it explicitly tests for definedness and then
+passes the result to the following block. Similarly, `without` tests that the object is undefined and also
+passes the result to the following block.
+
+For more natural control flow with undefined and defined types, Perl6 introduces `andthen` and `orelse`.
+
+    my $result = parse-int($string) orelse 0;    
+    
+    sub hello() { say 'hi' }
+    hello() andthen say 'bye';
+
+TODO: include a better example for andthen that makes sense. Maybe using promise objects?
+    
+So in practice, Perl6 does not have the concept of a null type, but rather of defined or undefined types.
+
+=head2 Data Definitions
+
+Perl6 is fundamentally an Object Oriented language. However, it also gives you the freedom
+to write in virtually any paradigm you wish. If you only want to pure functions that take an
+object and return a new object, you can certainly do so.
+
+Here is a Haskell code example:
+
+    data Point = Point x y
+    
+    moveUp :: Point -> Point
+    moveUp (Point x y) = Point x (y + 1)
+    
+And an equivalent Perl6 example:
+
+    class Point { has $.x; has $.y; }
+    
+    sub move-up(Point $p --> Point) {
+      Point.new(x => $p.x, y => $p.y + 1)
+    }
+
+The code I illustrated above is an example of a Product Type. If instead you'd like to
+write a Sum Type, there is not an exact equivalent in Perl6. The closest thing would be
+an Enum.
+
+    data Animal = Dog | Cat | Bird | Horse
+    
+    testAnimal :: Animal -> String
+    testAnimal Dog   = "Woof"
+    testAnimal Horse = "Ney"
+    
+A Perl6 Enum does not fit the same exact use cases, but it can be used in putting
+constraints on types.
+
+    enum Animal < Dog Cat Bird Horse >;
+
+    proto sub test-animal( Animal        ) {*}
+    multi sub test-animal( Dog           ) { 'Woof' }
+    multi sub test-animal( Animal::Horse ) { 'Ney'  }   # more explicit
+
+    say test-animal Animal::Dog; # more explicit
+    say test-animal Horse;
+
+=head2 Type Aliases and Subsets
+
+In Haskell, you can alias an existing type to simply increase clarity of intent and re-use
+existing types.
+
+    type Name = String
+    
+    fullName :: Name -> Name -> Name
+    fullName first last = first ++ last
+    
+The equivalent in Perl6 is the following.
+
+    my constant Name = Str;
+
+    sub full-name ( Name \first, Name \last --> Name ) { first ~ last }
+    
+It should be noted that in Perl6, one can also created a subset of an existing type.
+
+    subset Name of String where { *.chars < 20 };
```

Comment:
This should be `subset Name of Str where *.chars < 20;`

Also a typo in the line above, "created" => "create"?